### ISSUE TITLE
Smart Initialization Inference for TypeScript Variables

### ISSUE THREAD

**Issue by PierreTurnbull** [microsoft/TypeScript#56527]
### 🔍 Search Terms

"initialization", "variable", "inference", "assignment", "assigned", "automatic detection"

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

One's code may implicitely involve that all cases have been processed in an if-else structure. However TS does not detect it.

Consider the following example:

```ts
type Dinosaur = "triceratops" | "brachiosaurus"

const myCollectionOfDinosaurs: Dinosaur[] = ["triceratops", "brachiosaurus"]

const randomKey = Math.floor(Math.random() * myCollectionOfDinosaurs.length)
const randomlyChosenDinosaur: Dinosaur = myCollectionOfDinosaurs[randomKey]

let dinosaurGreeting: string

if (randomlyChosenDinosaur === "triceratops") dinosaurGreeting = "Hello triceratops!"
else if (randomlyChosenDinosaur === "brachiosaurus") dinosaurGreeting = "Hello brachiosaurus!"

console.log(dinosaurGreeting)
```

TS complains on the last line because it says dinosaurGreeting is used before being assigned.

It would be great if it automatically detected that the if else structure covers the whole type of Dinosaur. As a consequence in my example, it would infer that dinosaurGreeting has necessarily been assigned, and it would not complain in the console.log.

This would make sense as it would only be an improvement on an existing feature of TS: its ability to detect whether a variable is assigned before being used.

Consider the following example:

```ts
let a: string

if (true) a = "a"

console.log(a)
```

TS does not complain about a being used as it realizes it is necessarily assigned when logged.

### 📃 Motivating Example

I'd use the dinosaurs example above, and assert that this feature would remove unnecessary constraints in the language.

### 💻 Use Cases

1. What do you want to use this for?
Writing some meaningful code with less constraints, shaping it more precisely and remaining safe from type-related problems.
3. What shortcomings exist with current approaches?
Current approaches involve working around the way you intend your code to work. For example I can use !: when typing the variable, or initialize the variable with a default variable, or define it anywhere it is initialized, but all these solutions would imply thinking the code differently, in a way that is less compatible with my need.
5. What workarounds are you using in the meantime?
I cannot find the name for this operator : `!:`
```ts
let dinosaurGreeting!: string
```
This tells TS I know what I'm doing. Not exactly a solution, more a clunky workaround.

**Comment by Andarist** [comment]
To rephrase this a little bit - you are asking for incorporating exhaustiveness analysis into the control flow of if statements. It's worth noting that the same code rewritten to `switch`/`case` works just fine ([TS playground](https://www.typescriptlang.org/play?ts=5.3.2#code/C4TwDgpgBAIglgOwPYGcCGBXATlAvFAImCzgGMIs1gkwUCoAfQgI0tIAs5VMsM6BuALAAoEQBMIpADZos0UkgQpgUSgjFIAtlJABhdqggJ4ydNgBcsRN2xDRwqRBVjrZrAHE5TxAHNLykgQfOxEUAHc4YA4oAAo1DW09AxQjExssAEooAG8RKChSNBSWNk50vgJzPPyoF1MeTwhvILxCAAkIKSkkKFY0Di43PgBCAjsa3rk0AGtxgqLoIhJySmpaSur8uvTG5p9Wgg6unuIyCioaFFG5-NYIGbsAXxERBSUkRwA6bp8Y7bddsBfBk7EA)):
```ts
type Dinosaur = "triceratops" | "brachiosaurus";

declare const randomlyChosenDinosaur: Dinosaur;

let dinosaurGreeting: string;

switch (randomlyChosenDinosaur) {
  case "brachiosaurus":
    dinosaurGreeting = "Hello brachiosaurus!";
    break;
  case "triceratops":
    dinosaurGreeting = "Hello triceratops!";
    break;
}

console.log(dinosaurGreeting);
```

**Comment by PierreTurnbull** [comment]
Thank you @Andarist , the switch case is indeed the best solution to fix this problem, as it does not involve altering the logic of the code, only the syntax. I edited my post to add this solution.
